src/vm/hints/secp/secp_utils.rs

```diff
@@ -1,34 +1,48 @@
 use crate::bigint;
+use crate::math_utils::as_int;
 use crate::vm::errors::vm_errors::VirtualMachineError;
 use num_bigint::BigInt;
-use num_integer::Integer;
-use num_traits::{FromPrimitive, Signed};
-use std::ops::Shl;
+use num_traits::{FromPrimitive, Signed, Zero};
 
 /*
 Takes a 256-bit integer and returns its canonical representation as:
 d0 + BASE * d1 + BASE**2 * d2,
 where BASE = 2**86.
 */
-pub fn split(integer: &BigInt) -> Result<Vec<BigInt>, VirtualMachineError> {
+pub fn split(integer: &BigInt) -> Result<[BigInt; 3], VirtualMachineError> {
     if integer.is_negative() {
         return Err(VirtualMachineError::SecpSplitNegative(integer.clone()));
     }
-    let base = bigint!(1).shl(86);
+    let base = bigint!(1) << 86_usize;
+    let base_max = base - bigint!(1);
     let mut num = integer.clone();
-    let mut remainder: BigInt;
-    let mut canonical_repr = Vec::new();
-    for _n in 0..3 {
-        (num, remainder) = num.clone().div_rem(&base);
-        canonical_repr.push(remainder);
+    let mut canonical_repr: [BigInt; 3] = Default::default();
+    for item in &mut canonical_repr {
+        *item = (&num & &base_max).to_owned();
+        num >>= 86_usize;
     }
-    if num != bigint!(0) {
+    if !num.is_zero() {
         return Err(VirtualMachineError::SecpSplitutOfRange(integer.clone()));
     }
-
     Ok(canonical_repr)
 }
 
+/*
+Takes an UnreducedBigInt3 struct which represents a triple of limbs (d0, d1, d2) of field
+elements and reconstructs the corresponding 256-bit integer (see split()).
+Note that the limbs do not have to be in the range [0, BASE).
+prime should be the Cairo field, and it is used to handle negative values of the limbs.
+*/
+pub fn pack(d0: &BigInt, d1: &BigInt, d2: &BigInt, prime: &BigInt) -> BigInt {
+    let unreduced_big_int_3 = vec![d0, d1, d2];
+
+    unreduced_big_int_3
+        .iter()
+        .enumerate()
+        .map(|(idx, value)| as_int(value, prime) * (bigint!(1) << (idx * 86)))
```

Comment:
I missed a spot, product by the power of two is equivalent to making the shift only:
```suggestion
        .map(|(idx, value)| as_int(value, prime) << (idx * 86))
```